### PR TITLE
KAFKA-17331: Throw UnsupportedVersionException if the data in ListOffsetRequest does NOT fit EarliestLocalSpec and LatestTieredSpec.

### PR DESCRIPTION
**Pull request by frankvicky**
JIRA: [KAFKA-17331](https://issues.apache.org/jira/browse/KAFKA-17331)

Add the version check to client side when building `ListOffsetRequest`  for the specific timestamp:
- the version must be `>=8` if `timestamp=-4L` (`EARLIEST_LOCAL_TIMESTAMP`)
- the version must be `>=9` if `timestamp=-5L` (`LATEST_TIERED_TIMESTAMP`)

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java [hunk 1]
```diff
@@ -95,13 +96,26 @@ public Builder setTargetTimes(List<ListOffsetsTopic> topics) {
...
 
         @Override
         public String toString() {
             return data.toString();
         }
+
+        private void checkVersion(short version, ListOffsetsPartition partition) {
+            long timestamp = partition.timestamp();
+            if (timestamp == EARLIEST_LOCAL_TIMESTAMP && version < 8)
+                throw new UnsupportedVersionException("apiVersion must be >= 8 for EARLIEST_LOCAL_TIMESTAMP");
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java]: Could you please add the tp to the error message? that is helpful to users.

File: clients/src/test/java/org/apache/kafka/common/requests/ListOffsetsRequestTest.java [hunk 2]
```diff
@@ -146,4 +150,30 @@ public void testToListOffsetsTopics() {
         assertTrue(topic.partitions().contains(lop1));
     }
 
+    @Test
+    public void testCheckVersion() {
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/common/requests/ListOffsetsRequestTest.java]: Could you please separate them into 2 test cases? that is more readable.

File: clients/src/test/java/org/apache/kafka/common/requests/ListOffsetsRequestTest.java [hunk 3]
```diff
@@ -146,4 +150,30 @@ public void testToListOffsetsTopics() {
...
+        );
+
+        ListOffsetsRequest.Builder builder = ListOffsetsRequest.Builder
+                .forConsumer(true,
+                        IsolationLevel.READ_COMMITTED,
+                        false,
+                        false)
+                .setTargetTimes(topics);
+
+        assertThrows(UnsupportedVersionException.class, () -> builder.build(version));
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/common/requests/ListOffsetsRequestTest.java]: Please check the error message to make sure that is expected error.

**Review by chia7712 — COMMENTED** [review]
@frankvicky thanks for this patch

**Comment by frankvicky** [comment]
Hi @chia7712 
I have addressed the comments, PTAL

File: clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java [hunk 4]
```diff
@@ -95,13 +96,33 @@ public Builder setTargetTimes(List<ListOffsetsTopic> topics) {
 
         @Override
         public ListOffsetsRequest build(short version) {
+            for (ListOffsetsTopic topic : data.topics()) {
+                for (ListOffsetsPartition partition : topic.partitions()) {
+                    checkVersion(version, topic.name(), partition);
```
Comment (FrankYang0529) [inline clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java]: The failed case `ListOffsetsRequestTest#testResponseDefaultOffsetAndLeaderEpochForAllVersions` can be reproduced on my laptop. Could you take a look? Thank you.

**Review by FrankYang0529 — CHANGES_REQUESTED** [review]
Thanks for the PR. Leave one comment.

File: clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java (hunk 4 above)
Comment (frankvicky) [reply clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java]: Thank you, I will take a look!

Comment (chia7712) [reply clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java]: @frankvicky Could you please add a helper to build invalid request in order to run test?

Comment (frankvicky) [reply clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java]: Hi @FrankYang0529 and @chia7712 
I have reviewed the code. Since this error is due to a version check in the build method, I will change `assertEqual` to `assertThrows` when the test tries to build an invalid request to align with the current behavior.

Comment (chia7712) [reply clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java]: @frankvicky What we want to protect is the request from clients (admin and consumer). So maybe we should revise `fromConsumer`[0] rather than `build`.

WDYT?

[0] https://github.com/apache/kafka/blob/21dd5cd421139d6134d738d6e7555f0db6e1f622/clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java#L65

Comment (chia7712) [reply clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java]: The benefit is that we can leverage the API_VERSION check of network layer. The disadvantage is that it requires the API_VERSION, and fortunately that is existent in consumer and admin.

Comment (frankvicky) [reply clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java]: Hi @chia7712
I have reviewed the code. Since `forConsumer` is a static method, if we aim to perform a version check in this method, we will need to add a new argument. I will try to implement it.

File: clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java [hunk 5]
```diff
@@ -65,7 +66,8 @@ public static Builder forReplica(short allowedVersion, int replicaId) {
         public static Builder forConsumer(boolean requireTimestamp,
                                           IsolationLevel isolationLevel,
                                           boolean requireMaxTimestamp,
-                                          boolean requireTieredStorageTimestamp) {
+                                          boolean requireTieredStorageTimestamp,
+                                          List<ListOffsetsTopic> topics) {
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java]: Could you please follow the pattern? You can add a flag to pick up the min version. Current implementation does not consider the version=8 and I feel that is a kind of bug. @FrankYang0529 WDYT?

Comment (frankvicky) [reply clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java]: Hi @chia7712 
I have add a `requireEarliestLocalTimestamp` parameter to `forConsumer` method, PTAL

Comment (FrankYang0529) [reply clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java]: Yes, we should update this. Thanks @chia7712 and @frankvicky

File: clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java [hunk 6]
```diff
@@ -65,17 +66,25 @@ public static Builder forReplica(short allowedVersion, int replicaId) {
...
             else if (requireMaxTimestamp)
                 minVersion = 7;
             else if (isolationLevel == IsolationLevel.READ_COMMITTED)
                 minVersion = 2;
             else if (requireTimestamp)
                 minVersion = 1;
-            return new Builder(minVersion, ApiKeys.LIST_OFFSETS.latestVersion(), CONSUMER_REPLICA_ID, isolationLevel);
+
+            short version = ApiKeys.LIST_OFFSETS.latestVersion();
+            topics.forEach(topic -> topic.partitions().forEach(partition -> checkVersion(version, topic.name(), partition)));
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java]: After setting the min version, the version checkcan handled by `NetworkClient` [0]. Hence, we don't need to add extra check here.

[0] https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/NetworkClient.java#L528

File: clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java [hunk 7]
```diff
@@ -65,17 +66,25 @@ public static Builder forReplica(short allowedVersion, int replicaId) {
         public static Builder forConsumer(boolean requireTimestamp,
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java:66]: There are too many boolean arguments. Could you please refactor it? for instance:
```java
        public static Builder forTieredStorageTimestamp(IsolationLevel isolationLevel) {

            return new Builder((short) 9, ApiKeys.LIST_OFFSETS.latestVersion(), CONSUMER_REPLICA_ID, isolationLevel);
        }
```

**Review by chia7712 — COMMENTED** [review]
@frankvicky thanks for this patch

File: clients/src/main/java/org/apache/kafka/clients/admin/internals/ListOffsetsHandler.java [hunk 8]
```diff
@@ -93,13 +94,38 @@ ListOffsetsRequest.Builder buildBatchedRequest(int brokerId, Set<TopicPartition>
             .stream()
             .anyMatch(key -> offsetTimestampsByPartition.get(key) == ListOffsetsRequest.MAX_TIMESTAMP);
 
+        boolean requireEarliestLocalTimestamp = keys
+                .stream()
+                .anyMatch(key -> offsetTimestampsByPartition.get(key) == ListOffsetsRequest.EARLIEST_LOCAL_TIMESTAMP);
+
         boolean requireTieredStorageTimestamp = keys
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/admin/internals/ListOffsetsHandler.java:100]: Could you please move this "helper" to `ListOffset.Builder`? normally, we should not use this helper as it requires to "loop" data. However, if callers have no idea about the "timestamp" type, this helper can find out that for them.

Comment (frankvicky) [reply clients/src/main/java/org/apache/kafka/clients/admin/internals/ListOffsetsHandler.java:100]: Sounds good 😃

**Comment by chia7712** [comment]
@frankvicky I feel the refactor is a bit overkill. Maybe we can add following two methods:

```java
        public static Builder forConsumer(boolean requireTimestamp,
                                          IsolationLevel isolationLevel) {
            return forConsumer(requireTimestamp, isolationLevel, false, false, false);
        }

        public static Builder forConsumer(boolean requireTimestamp,
                                          IsolationLevel isolationLevel,
                                          boolean requireMaxTimestamp,
                                          boolean requireEarliestTimestamp,
                                          boolean requireTieredStorageTimestamp) {
            short minVersion = 0;
            if (requireTieredStorageTimestamp)
                minVersion = 9;
            else if (requireEarliestTimestamp)
                minVersion = 8;
            else if (requireMaxTimestamp)
                minVersion = 7;
            else if (isolationLevel == IsolationLevel.READ_COMMITTED)
                minVersion = 2;
            else if (requireTimestamp)
                minVersion = 1;
            return new Builder(minVersion, ApiKeys.LIST_OFFSETS.latestVersion(), CONSUMER_REPLICA_ID, isolationLevel);
        }
```
Those two methods should cover all use cases for now.

**Comment by frankvicky** [comment]
Hi @chia7712 
I have refactor the `forConsumer` method, PTAL

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsRequestManager.java [hunk 9]
```diff
@@ -336,9 +336,8 @@ private CompletableFuture<ListOffsetResult> buildListOffsetRequestToNode(
             Map<TopicPartition, ListOffsetsRequestData.ListOffsetsPartition> targetTimes,
             boolean requireTimestamps,
             List<NetworkClientDelegate.UnsentRequest> unsentRequests) {
-        ListOffsetsRequest.Builder builder = ListOffsetsRequest.Builder
-                .forConsumer(requireTimestamps, isolationLevel, false, false)
-                .setTargetTimes(ListOffsetsRequest.toListOffsetsTopics(targetTimes));
+        List<ListOffsetsRequestData.ListOffsetsTopic> topics = ListOffsetsRequest.toListOffsetsTopics(targetTimes);
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsRequestManager.java]: If there is no strong reason, could we keep fluent pattern? 
```java
        ListOffsetsRequest.Builder builder = ListOffsetsRequest.Builder
                .forConsumer(requireTimestamps, isolationLevel)
                .setTargetTimes(ListOffsetsRequest.toListOffsetsTopics(targetTimes));
```

File: clients/src/test/java/org/apache/kafka/common/requests/ListOffsetsRequestTest.java [hunk 10]
```diff
@@ -68,11 +76,11 @@ public void testGetErrorResponse() {
                                 new ListOffsetsPartition()
                                     .setPartitionIndex(0))));
             ListOffsetsRequest request = ListOffsetsRequest.Builder
-                    .forConsumer(true, IsolationLevel.READ_COMMITTED, false, false)
+                    .forConsumer(true, IsolationLevel.READ_COMMITTED)
                     .setTargetTimes(topics)
                     .build(version);
             ListOffsetsResponse response = (ListOffsetsResponse) request.getErrorResponse(0, Errors.NOT_LEADER_OR_FOLLOWER.exception());
-
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/common/requests/ListOffsetsRequestTest.java]: please revert this unrelated change

File: clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java [hunk 11]
```diff
@@ -2354,9 +2354,10 @@ private ListOffsetsRequest createListOffsetRequest(short version) {
                             .setTimestamp(1000000L)
                             .setMaxNumOffsets(10)
                             .setCurrentLeaderEpoch(5)));
+            List<ListOffsetsTopic> topics = singletonList(topic);
             return ListOffsetsRequest.Builder
-                    .forConsumer(false, IsolationLevel.READ_UNCOMMITTED, false, false)
-                    .setTargetTimes(singletonList(topic))
+                    .forConsumer(false, IsolationLevel.READ_UNCOMMITTED)
+                    .setTargetTimes(topics)
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java]: ditto

File: clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java [hunk 12]
```diff
@@ -2365,9 +2366,10 @@ private ListOffsetsRequest createListOffsetRequest(short version) {
                             .setPartitionIndex(0)
                             .setTimestamp(1000000L)
                             .setCurrentLeaderEpoch(5)));
+            List<ListOffsetsTopic> topics = singletonList(topic);
             return ListOffsetsRequest.Builder
-                    .forConsumer(true, IsolationLevel.READ_UNCOMMITTED, false, false)
-                    .setTargetTimes(singletonList(topic))
+                    .forConsumer(true, IsolationLevel.READ_UNCOMMITTED)
+                    .setTargetTimes(topics)
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java]: ditto

File: clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java [hunk 13]
```diff
@@ -2378,9 +2380,10 @@ private ListOffsetsRequest createListOffsetRequest(short version) {
             ListOffsetsTopic topic = new ListOffsetsTopic()
                     .setName("test")
                     .setPartitions(singletonList(partition));
+            List<ListOffsetsTopic> topics = singletonList(topic);
             return ListOffsetsRequest.Builder
-                    .forConsumer(true, IsolationLevel.READ_COMMITTED, false, false)
-                    .setTargetTimes(singletonList(topic))
+                    .forConsumer(true, IsolationLevel.READ_COMMITTED)
+                    .setTargetTimes(topics)
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java]: ditto

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala [hunk 14]
```diff
@@ -287,15 +287,15 @@ class AuthorizerIntegrationTest extends AbstractAuthorizerIntegrationTest {
...
-    requests.ListOffsetsRequest.Builder.forConsumer(false, IsolationLevel.READ_UNCOMMITTED, false, false).setTargetTimes(
-      List(new ListOffsetsTopic()
-        .setName(tp.topic)
-        .setPartitions(List(new ListOffsetsPartition()
-          .setPartitionIndex(tp.partition)
-          .setTimestamp(0L)
-          .setCurrentLeaderEpoch(27)).asJava)).asJava
-      ).
-      build()
+    val topics = List(new ListOffsetsTopic()
```
Comment (chia7712) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: ditto

File: core/src/test/scala/unit/kafka/server/LogOffsetTest.scala [hunk 15]
```diff
@@ -60,8 +60,9 @@ class LogOffsetTest extends BaseRequestTest {
   @ValueSource(strings = Array("zk", "kraft"))
   def testGetOffsetsForUnknownTopic(quorum: String): Unit = {
     val topicPartition = new TopicPartition("foo", 0)
-    val request = ListOffsetsRequest.Builder.forConsumer(false, IsolationLevel.READ_UNCOMMITTED, false, false)
-      .setTargetTimes(buildTargetTimes(topicPartition, ListOffsetsRequest.LATEST_TIMESTAMP, 10).asJava).build(0)
+    val targetTimes = buildTargetTimes(topicPartition, ListOffsetsRequest.LATEST_TIMESTAMP, 10).asJava
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/server/LogOffsetTest.scala]: ditto

File: core/src/test/scala/unit/kafka/server/RequestQuotaTest.scala [hunk 16]
```diff
@@ -288,8 +288,8 @@ class RequestQuotaTest extends BaseRequestTest {
               .setPartitionIndex(tp.partition)
               .setTimestamp(0L)
               .setCurrentLeaderEpoch(15)).asJava)
-          ListOffsetsRequest.Builder.forConsumer(false, IsolationLevel.READ_UNCOMMITTED, false, false)
-            .setTargetTimes(List(topic).asJava)
+          val targetTimes = List(topic).asJava
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/server/RequestQuotaTest.scala]: ditto

File: clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java [hunk 17]
```diff
@@ -102,6 +110,7 @@ public ListOffsetsRequest build(short version) {
         public String toString() {
             return data.toString();
         }
+
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java]: ditto

**Review by chia7712 — COMMENTED** [review]
@frankvicky thanks for this patch. a couple of comments are left.

**Review by chia7712 — APPROVED** [review]
LGTM

File: clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java [hunk 18]
```diff
@@ -58,17 +58,21 @@ public class ListOffsetsRequest extends AbstractRequest {
...
 
         public static Builder forConsumer(boolean requireTimestamp,
                                           IsolationLevel isolationLevel,
                                           boolean requireMaxTimestamp,
+                                          boolean requireEarliestTimestamp,
                                           boolean requireTieredStorageTimestamp) {
             short minVersion = 0;
             if (requireTieredStorageTimestamp)
                 minVersion = 9;
+            else if (requireEarliestTimestamp)
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java]: @FrankYang0529 Could you please take a look?

Comment (FrankYang0529) [reply clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java]: Hi @chia7712 and @frankvicky, What does `requireEarliestTimestamp` mean? `EARLIEST_TIMESTAMP` or `EARLIEST_LOCAL_TIMESTAMP`? If it's `EARLIEST_LOCAL_TIMESTAMP`, probably we should change the parameter name as `requireEarliestLocalTimestamp`, so users don't misunderstand it.

Comment (frankvicky) [reply clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java]: Sure, I will update it

**Comment by chia7712** [comment]
@FrankYang0529 Could you please take a look?

**Review by FrankYang0529 — APPROVED** [review]
LGTM. Thanks for the fix.
